### ISSUE TITLE
nx migrate is not working at Yarn PnP

### ISSUE THREAD

**Issue by enk0de** [nrwl/nx#20681]
### Current Behavior

as i want to migrate v16 to v17, i did `yarn nx migrate latest`

but i confronted below error
```
/bin/sh: /var/folders/dl/mvnz7vgs14z7zylgz_zlpwt40000gn/T/tmp-82907-i6iOTIpmDoDh/node_modules/.bin/nx: No such file or directory
Command failed: /var/folders/dl/mvnz7vgs14z7zylgz_zlpwt40000gn/T/tmp-82907-i6iOTIpmDoDh/node_modules/.bin/nx _migrate latest
```

### Expected Behavior

run migration correctly

### Steps to Reproduce

1. using yarn 3+(or 4) with PnP in nx workspace
2. run `nx migrate latest`

### Nx Report

```shell
-
```

### Package Manager Version

4.0.1

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by enk0de** [comment]
yes. here is `nx report` result

```
>  NX  Falling back to ts-node for local typescript execution. This may be a little slower.
  - To fix this, ensure @swc-node/register and @swc/core have been installed

 >  NX   Report complete - copy this into the issue template

   Node   : 20.6.1
   OS     : darwin-arm64
   yarn   : 3.6.3
   
   nx                 : 16.8.1
   @nx/js             : 16.8.1
   @nx/jest           : 16.8.1
   @nx/linter         : 16.8.1
   @nx/workspace      : 16.8.1
   @nx/cypress        : 16.8.1
   @nx/eslint-plugin  : 16.8.1
   @nx/next           : 16.8.1
   @nx/node           : 16.8.1
   @nx/react          : 16.8.1
   @nx/rollup         : 16.8.1
   @nx/web            : 16.8.1
   typescript         : 5.1.6
   ---------------------------------------
   Community plugins:
   @jscutlery/semver : 2.29.3


```

**Comment by meeroslav** [comment]
Hm... I could not reproduce this error, but I got another one

![Screenshot 2023-12-11 at 14 42 46](https://github.com/nrwl/nx/assets/881612/ae817258-e0a9-472e-8e63-91f07ce9a204)

Also, the initial migration only worked for me if I had `v8-compile-cache` installed explicitly (this is dependency of nx 17.1.1 but not 17.2.0).
![Screenshot 2023-12-11 at 14 43 58](https://github.com/nrwl/nx/assets/881612/655d07ca-d8e8-4679-bce1-9dc5a33e9af2)

@AgentEnder, can you look into this?

Use this as a starting point: https://github.com/meeroslav/nx-20681

**Comment by taylor-schley-kr** [comment]
We have run into this issue when yarn nodeLinker was set for node-modules in the calling package. It looks as thought the tmp package is being generated with the nodeLinker not set, so the tmp package is using yarn PnP, but the migrate is expecting node-modules.

**Comment by Puertas** [comment]
We also are run into this when trying to migrate to v18, also with Yarn, tried with both 4.0.2 and 4.1.1 but still same error:

```
/bin/sh: /var/folders/r3/r_c9w7ys7qgf1n8jj36s4_9w0000gn/T/tmp-15567-Q7Fe3yA67exQ/node_modules/.bin/nx: No such file or directory
Command failed: /var/folders/r3/r_c9w7ys7qgf1n8jj36s4_9w0000gn/T/tmp-15567-Q7Fe3yA67exQ/node_modules/.bin/nx _migrate latest
``` 

The report just in case that helps:

```
>  NX   Report complete - copy this into the issue template

   Node   : 20.11.0
   OS     : darwin-arm64
   yarn   : 4.1.1

   nx                 : 17.2.8
   @nx/js             : 17.2.8
   @nx/jest           : 17.2.8
   @nx/linter         : 17.2.8
   @nx/eslint         : 17.2.8
   @nx/workspace      : 17.2.8
   @nx/cypress        : 17.2.8
   @nx/devkit         : 17.2.8
   @nx/eslint-plugin  : 17.2.8
   @nx/express        : 17.2.8
   @nx/next           : 17.2.8
   @nx/node           : 17.2.8
   @nx/react          : 17.2.8
   @nx/storybook      : 17.2.8
   @nrwl/tao          : 17.2.8
   @nx/web            : 17.2.8
   @nx/webpack        : 17.2.8
   nx-cloud           : 16.5.2
   typescript         : 5.2.2
```

**Comment by FrozenPandaz** [comment]
I believe this may have been fixed. Could you please retry with the latest version?

**Comment by Puertas** [comment]
For us it's fixed, I did a try with last version (19.0.3) and working fine with yarn
